syntax = "proto3";

package raft;

service Raft {
    rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse) {};
    rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse) {};

    // example api
    rpc SayHello(HelloRequest) returns (HelloReply) {};
    rpc SayHelloStreamReply(HelloRequest) returns (stream HelloReply) {};
}

message RequestVoteRequest {
    uint64 term = 1;
    uint64 candidate_id = 2;
    uint64 last_log_index = 3;
    uint64 last_log_term = 4;
}

message RequestVoteResponse {
    uint64 term = 1;
    bool vote_granted = 2;
}

message LogEntry {
    uint64 index = 1;
    uint64 term = 2;
    string command = 3;
}

message AppendEntriesRequest {
    uint64 term = 1;
    uint64 leader_id = 2;
    uint64 prev_log_index = 3;
    uint64 prev_log_term = 4;
    repeated LogEntry entries = 5;
    uint64 leader_commit = 6;
}

message AppendEntriesResponse {
    uint64 term = 1;
    bool success = 2;
}

message HelloRequest {
    string name = 1;
}

message HelloReply {
    string message = 1;
}